src/Protocols/PacketStream/AsyncFifo.hs

```diff
@@ -0,0 +1,60 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+
+{- |
+Provides `asyncFifoC` for crossing clock domains in the packet stream protocol.
+-}
+module Protocols.PacketStream.AsyncFifo (asyncFifoC) where
+
+import Data.Maybe.Extra (toMaybe)
+
+import Clash.Explicit.Prelude (asyncFIFOSynchronizer)
+import Clash.Prelude
+
+import Protocols.Internal (Circuit, fromSignals)
+import Protocols.PacketStream.Base
+
+{- | Asynchronous FIFO circuit that can be used to safely cross clock domains.
+Uses `Clash.Explicit.Prelude.asyncFIFOSynchronizer` internally.
+-}
+asyncFifoC ::
+  forall
+    (depth :: Nat)
+    (dataWidth :: Nat)
+    (wDom :: Domain)
+    (rDom :: Domain)
+    (metaType :: Type).
+  ( KnownDomain wDom
+  , KnownDomain rDom
+  , KnownNat depth
+  , 2 <= depth
+  , KnownNat dataWidth
+  , 1 <= dataWidth
+  , NFDataX metaType
+  ) =>
+  -- | 2^depth is the number of elements this component can store
+  SNat depth ->
+  -- | Clock signal in the write domain
+  Clock wDom ->
+  -- | Reset signal in the write domain
+  Reset wDom ->
+  -- | Enable signal in the write domain
+  Enable wDom ->
+  -- | Clock signal in the read domain
+  Clock rDom ->
+  -- | Reset signal in the read domain
+  Reset rDom ->
+  -- | Enable signal in the read domain
+  Enable rDom ->
+  Circuit (PacketStream wDom dataWidth metaType) (PacketStream rDom dataWidth metaType)
+asyncFifoC depth wClk wRst wEn rClk rRst rEn = fromSignals ckt
+ where
+  ckt (fwdIn, bwdIn) = (bwdOut, fwdOut)
+   where
+    (element, isEmpty, isFull) = asyncFIFOSynchronizer depth wClk rClk wRst rRst wEn rEn readReq fwdIn
+    notEmpty = not <$> isEmpty
+    -- If the FIFO is empty, we output Nothing. Else, we output the oldest element.
+    fwdOut = toMaybe <$> notEmpty <*> element
+    -- Assert backpressure when the FIFO is full.
+    bwdOut = PacketStreamS2M . not <$> isFull
```

Comment:
Data will be lost when the fifo is held in reset. Perhaps `asyncFifoSynchronizer` should be changed so that it asserts `full` when the write side is in reset.

Reply:
True. `forceResetSanity` will fix that as well.